### DISCUSSION TITLE
Answer to print(9 +-+-+-+ 2) in Python?

### DISCUSSION THREAD

**Discussion by Vajeri** [community/community#122966] [Programming Help] [answered]
### Body

why does print(9 +-+-+-+ 2) give an answer of 7 in python?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CopperEagle** [comment, marked as answer]
Hey @Vajeri , funky question :)

In the `+-+-+-+` string, the leftmost is treated as binary operator (taking two arguments) and the others modify the sign of 2.

Starting with a simpler example:
```py
a = 9 +- 2
a == 9 + (-2) == 7
```

Then, getting the result of the expression you provide is like solving one of these algebra exercises from school: 
```py
a = 9 +-+-+-+ 2
  = 9 + (- (+ (- (+ (- (+ 2))))))
  = 9 (- (+ (- (+ (- 2)))))
  = 9 (- (+ (- (- 2))))
  = 9 (- (+ 2))
  = 9 - 2 == 7
```

If this answers your question, please consider marking it as the answer :)

**Comment by spenserblack** [comment]
Let's break it down a bit:
```python
print(9 +- 7)
```
The spacing makes it confusing, but it's `9` plus negative 7 (`-7`).
```python
print(9 + -7)
```

Now the interesting thing is that `+n` and `-n` are [unary operators](https://docs.python.org/3/reference/expressions.html#unary-arithmetic-and-bitwise-operations). It's the same as calling `n.__pos__()` and `n.__neg__()`. So `9 + -7` is the same as
```python
9 + (2).__neg__()
```

When the `-2` is lead by an extra `+`, you're calling the *positive* unary operator on the result of `-2`, so you get
```python
9 + (2).__neg__().__pos__()
```

And all the other `-+` are just extending that chain.

Also, even addition is a function call.

```python
(9).__add__((2).__neg__().__pos__())
```

So, you can write what you have as this to make it a bit more readable:

```python
left = 9
right = -+-+-+ 2
print(left + right)  # 7
```

Here's an experiment:

```python
class MyInt(int):
    """
    This class acts like an int, except it will print the function that gets called.
    """
    def __add__(self, other):
        print('add!')
        return MyInt(super().__add__(other))

    def __neg__(self):
        print('neg!')
        return MyInt(super().__neg__())

    def __pos__(self):
        print('pos!')
        return MyInt(super().__pos__())

print(MyInt(9) +-+-+-+ MyInt(2))
```
